### PR TITLE
Add checks feature flag

### PR DESCRIPTION
**Pull request by amitaekbote**
Add a flag to enable checks for dcos during upgrades

### CODE REVIEW

**Comment by amitaekbote** [comment]
@mellenburg ready for review, https://github.com/dcos/dcos/pull/1649 works fine with the change

File: dcos_test_utils/upgrade.py [hunk 1]
```diff
@@ -125,17 +127,30 @@ def upgrade_dcos(
...
-            try:
-                wait_for_mesos_metric(dcos_api_session, host, wait_metric)
-            except retrying.RetryError as exc:
-                raise Exception(
-                    'Timed out waiting for {} to rejoin the cluster after upgrade: {}'.
-                    format(role_name, repr(host))
-                ) from exc
+            if use_checks:
+                try:
+                    output = ssh_client.command(host, ['sudo', 'bash', 'dcos_node_upgrade.sh'], stdout="")
```
Comment (mellenburg) [inline dcos_test_utils/upgrade.py]: Why are you assigning stdout="" ?

File: dcos_test_utils/upgrade.py [hunk 2]
```diff
@@ -125,17 +127,30 @@ def upgrade_dcos(
...
+                    log.info("Output from checks" + output)
+                except CalledProcessError as e:
+                    returncode = e.returncode
+                    if returncode != 0:
+                        raise Exception(
+                            'Timed out waiting for {} to rejoin the cluster after upgrade: {}'.
+                            format(role_name, repr(host)))
+                    else:
+                        raise Exception("Dont know whats going on here")
```
Comment (mellenburg) [inline dcos_test_utils/upgrade.py]: Please do not do this: You are remapping an exception to nothing. Doing nothing at all would expose the actual error whereas this explicitly obfuscates it

File: dcos_test_utils/upgrade.py [hunk 3]
```diff
@@ -125,17 +127,30 @@ def upgrade_dcos(
...
-                    'Timed out waiting for {} to rejoin the cluster after upgrade: {}'.
-                    format(role_name, repr(host))
-                ) from exc
+            if use_checks:
+                try:
+                    output = ssh_client.command(host, ['sudo', 'bash', 'dcos_node_upgrade.sh'], stdout="")
+                    log.info("Output from checks" + output)
+                except CalledProcessError as e:
+                    returncode = e.returncode
+                    if returncode != 0:
```
Comment (mellenburg) [inline dcos_test_utils/upgrade.py]: If a CalledProcessError is raised then the error code MUST not be zero

File: dcos_test_utils/upgrade.py [hunk 4]
```diff
@@ -125,17 +127,30 @@ def upgrade_dcos(
...
-            except retrying.RetryError as exc:
-                raise Exception(
-                    'Timed out waiting for {} to rejoin the cluster after upgrade: {}'.
-                    format(role_name, repr(host))
-                ) from exc
+            if use_checks:
+                try:
+                    output = ssh_client.command(host, ['sudo', 'bash', 'dcos_node_upgrade.sh'], stdout="")
+                    log.info("Output from checks" + output)
+                except CalledProcessError as e:
```
Comment (mellenburg) [inline dcos_test_utils/upgrade.py]: Why do we need to catch this at all?

File: dcos_test_utils/upgrade.py [hunk 5]
```diff
@@ -125,17 +127,30 @@ def upgrade_dcos(
...
+                    returncode = e.returncode
+                    if returncode != 0:
+                        raise Exception(
+                            'Timed out waiting for {} to rejoin the cluster after upgrade: {}'.
```
Comment (mellenburg) [inline dcos_test_utils/upgrade.py]: I don't see what makes this a timeout error specifically. How is nonzero guaranteed to be a timeout? Is there no other failure case?

**Review by mellenburg — CHANGES_REQUESTED** [review]

**Comment by amitaekbote** [comment]
"output": "",
      "status": 0
    }
  }
}
Node checks passed

[2017-06-13 23:20:35,755|dcos_test_utils.helpers|DEBUG]: Request method GET: http://52.41.209.145/. Arguments: {}
[2017-06-13 23:20:35,757|urllib3.connectionpool|DEBUG]: Starting new HTTP connection (1): 52.41.209.145
[2017-06-13 23:20:35,816|urllib3.connectionpool|DEBUG]: http://52.41.209.145:80 "GET / HTTP/1.1" 200 33526
[2017-06-13 23:20:35,845|dcos_test_utils.dcos_api_session|INFO]: Nginx is UP!
[2017-06-13 23:20:37,106|urllib3.connectionpool|DEBUG]: http://52.41.209.145:80 "GET /agent/93f3c592-487b-4fcd-8405-388b6919909e-S1/files/download?path=%2Fvar%2Flib%2Fmesos%2Fslave%2Fslaves%2F93f3c592-487b-4fcd-8405-388b6919909e-S1%2Fframeworks%2F93f3c592-487b-4fcd-8405-388b6919909e-0000%2Fexecutors%2Fupgrade-dns-31d1471409bc40939ec95ccf6b5f1b52.154a63d3-50c7-11e7-a399-3a12cc4b4ffe%2Fruns%2Fe323cb8b-54b4-4038-a39f-386eb5078ba2%2Fdns_resolve_log.txt HTTP/1.1" 200 32767
xfail

============== 1 passed, 1 xfailed, 1 xpassed in 2217.72 seconds ===============
___________________________________ summary ____________________________________
  py35-upgrade-test: commands succeeded
  congratulations :)

**Comment by amitaekbote** [comment]
@mellenburg ready for another pass, works with flag in true or false state

File: dcos_test_utils/upgrade.py [hunk 6]
```diff
@@ -125,17 +127,26 @@ def upgrade_dcos(
...
-                    'Timed out waiting for {} to rejoin the cluster after upgrade: {}'.
-                    format(role_name, repr(host))
-                ) from exc
+            if use_checks == 'true':
+                try:
+                    output = ssh_client.command(host, ['sudo', 'bash', 'dcos_node_upgrade.sh'])
+                    log.info("Output from checks" + output.decode("utf-8"))
+                except CalledProcessError as e:
+                    raise Exception(
+                        'Checks didnt pass, {} didnt rejoin the cluster after upgrade: {}'.
```
Comment (adamtheturtle) [inline dcos_test_utils/upgrade.py]: `s/didnt/didn't/g` (or `did not`).

File: advanced_tests/test_upgrade.py [hunk 7]
```diff
@@ -1,6 +1,6 @@
 """
 
-WARNGING: This is a destructive process and you cannot go back
+WARNING: This is a destructive process and you cannot go back
 
 =========================================================
 This test has 4 input parameters (set in the environment)
```
Comment (adamtheturtle) [inline advanced_tests/test_upgrade.py]: Is this title still correct? (maybe it was incorrect before?)

**Comment by amitaekbote** [comment]
@adamtheturtle updated

File: advanced_tests/test_upgrade.py [hunk 8]
```diff
@@ -1,14 +1,16 @@
 """
 
-WARNGING: This is a destructive process and you cannot go back
```
Comment (mellenburg) [inline advanced_tests/test_upgrade.py:3]: I is a much good typer

File: advanced_tests/test_upgrade.py [hunk 9]
```diff
@@ -376,6 +379,9 @@ def upgraded_dcos(dcos_api_session, launcher, setup_workload, onprem_cluster, is
 
 
 @pytest.mark.skipif(
+    'TEST_UPGRADE_USE_CHECKS' not in os.environ,
+    reason='TEST_UPGRADE_USE_CHECKS must be set in env to upgrade a cluster')
```
Comment (mellenburg) [inline advanced_tests/test_upgrade.py:383]: Nit: maybe default this to false, but better to be explicit than not so

File: dcos_test_utils/upgrade.py [hunk 10]
```diff
@@ -43,7 +44,8 @@ def upgrade_dcos(
         dcos_api_session: dcos_test_utils.dcos_api_session.DcosApiSession,
         onprem_cluster: dcos_test_utils.onprem.OnpremCluster,
         starting_version: str,
-        installer_url: str) -> None:
+        installer_url: str,
+        use_checks: str) -> None:
```
Comment (mellenburg) [inline dcos_test_utils/upgrade.py]: Use a bool please

File: dcos_test_utils/upgrade.py [hunk 11]
```diff
@@ -125,17 +127,26 @@ def upgrade_dcos(
...
-            }[role]
-            log.info('Waiting for {} to rejoin the cluster...'.format(role_name))
-            try:
-                wait_for_mesos_metric(dcos_api_session, host, wait_metric)
-            except retrying.RetryError as exc:
-                raise Exception(
-                    'Timed out waiting for {} to rejoin the cluster after upgrade: {}'.
-                    format(role_name, repr(host))
-                ) from exc
+            if use_checks == 'true':
```
Comment (mellenburg) [inline dcos_test_utils/upgrade.py]: `if use_checks:`

File: dcos_test_utils/upgrade.py [hunk 12]
```diff
@@ -125,17 +127,26 @@ def upgrade_dcos(
...
+                try:
+                    output = ssh_client.command(host, ['sudo', 'bash', 'dcos_node_upgrade.sh'])
+                    log.info("Output from checks" + output.decode("utf-8"))
+                except CalledProcessError as e:
```
Comment (mellenburg) [inline dcos_test_utils/upgrade.py]: There is no reason to catch a CalledProcessed error in this manner. Also, it is rarely advisable to raise a new exception from a different exception without using `raise from`. If your goal is to make sure you have access to this output, I would suggest the following:
```
log.info("Starting upgrade script on {host} ({role_name})...".format(host=host, role_name=role_name))
if use_checks:
    ssh_client.command(host, ['sudo', 'bash', 'dcos_node_upgrade.sh'], stdout=sys.stdout.buffer)
else:
    ssh_client. ... [keep it the same hereafter]
```

This is much more robust as you are going to get the stdout on the driver machine instantly whereas the current implementation only works when the process actually finishes (which is far from guaranteed in testing)

File: advanced_tests/test_upgrade.py [hunk 13]
```diff
@@ -363,7 +365,8 @@ def upgraded_dcos(dcos_api_session, launcher, setup_workload, onprem_cluster, is
         upgrade_session,
         onprem_cluster,
         dcos_api_session.get_version(),
-        os.environ['TEST_UPGRADE_INSTALLER_URL'])
+        os.environ['TEST_UPGRADE_INSTALLER_URL'],
+        os.environ.get('TEST_UPGRADE_USE_CHECKS'))
```
Comment (mellenburg) [inline advanced_tests/test_upgrade.py]: This test is skipped if TEST_UPGRADE_USE_CHECKS is not in environ and we want this argument to be a bool so please do:
```
os.environ['TEST_UPGRADE_USE_CHECKS'] == 'true'
```

**Review by mellenburg — CHANGES_REQUESTED** [review]
Fixing the invocation of the ssh_client to run the upgrade script is critical for collecting logs.

File: dcos_test_utils/upgrade.py [hunk 14]
```diff
@@ -8,6 +8,7 @@
 import dcos_test_utils.onprem
 
 from dcos_test_utils import ssh_client
+from subprocess import CalledProcessError
```
Comment (mellenburg) [inline dcos_test_utils/upgrade.py]: please remove

File: dcos_test_utils/upgrade.py [hunk 15]
```diff
@@ -125,17 +127,25 @@ def upgrade_dcos(
...
-                wait_for_mesos_metric(dcos_api_session, host, wait_metric)
-            except retrying.RetryError as exc:
-                raise Exception(
-                    'Timed out waiting for {} to rejoin the cluster after upgrade: {}'.
-                    format(role_name, repr(host))
-                ) from exc
+            if use_checks:
+                try:
+                    ssh_client.command(host, ['sudo', 'bash', 'dcos_node_upgrade.sh'], stdout=sys.stdout.buffer)
+                except CalledProcessError as e:
```
Comment (mellenburg) [inline dcos_test_utils/upgrade.py]: please remove

**Review by mellenburg — CHANGES_REQUESTED** [review]
remapping the error is not good, so please just remove that entire try/except clause. If you want to know which  host+role was just triggered, then use that log statement I suggested before calling the SSH client

**Comment by amitaekbote** [comment]
The upgrade test failed because of a bug during teardown, actual test is 💯

**Review by mellenburg — APPROVED** [review]

**Comment by mellenburg** [comment]
Confirmed that failure is in fact due to a bug in the build configuration in CI, not this change at all. This is safe to promote to DC/OS testing now.

**Pull request merged**